Fix Upgrade docs content issues (#4943)
mattermost/docs: source/upgrade/upgrading-mattermost-server.rst

If you are upgrading from version 3.0 or later, these instructions apply to you. If you are upgrading from a version prior to 3.0.0, you must first upgrade to version 3.0.3.

> We only support one minor version difference between the server versions when performing a rolling upgrade (for example v5.27.1 + v5.27.2 or v5.26.4 + v5.27.1 is supported, whereas v5.25.5 + v5.27.0 is not supported). Running two different versions of Mattermost in your cluster should not be done outside of an upgrade scenario.

Author: @wiersgallak - This is the information we need to confirm as correct or update if needed. Is this still true? A customer has asked for clarification: "it's not clear for me whether I can upgrade from the 5.31 ESR to the 5.37 ESR, skipping everything in between or not.  The note at the end mentions minor versions and rolling releases - as opposed to what?"

Author: @wiersgallak I've clarified all upgrade path details under "Upgrading to the Latest Version".

  - Review the :doc:`important-upgrade-notes` to make sure you are aware of any actions you need to take before or after upgrading from your particular version.
  - If you're upgrading from a version prior to v5.0 be sure to also modify your service file to work with the binary changes introduced with 5.0. Your execution directory should point to the Mattermost base directory (i.e. ``/opt/mattermost``) and your binary should point to the ``mattermost`` binary (i.e. ``/opt/mattermost/bin/mattermost``).
  - Gather the following information before starting the upgrade:

>       - **Existing install directory - {install-path}**: If you don't know where Mattermost Server is installed, use the ``whereis mattermost`` command. The output should be similar to ``/opt/mattermost/bin/mattermost``. The install directory is everything before the first occurrence of the string ``/mattermost``. In this example, the ``{install-path}`` is ``/opt``. If that command does not produce any results because your version is older, try ``whereis platform`` instead.

Reviewer: ```suggestion
      - **Existing install directory - {install-path}**: If you don't know where Mattermost Server is installed, use the ``whereis mattermost`` command. The output should be similar to ``/opt/mattermost/bin/mattermost``. The install directory is everything before the first occurrence of the string ``/mattermost``. In this example, the ``{install-path}`` is ``/opt``. If that command does not produce any results it's likely because your version is older; try ``whereis platform`` instead.
```

>       - **Existing install directory - {install-path}**: If you don't know where Mattermost Server is installed, use the ``whereis mattermost`` command. The output should be similar to ``/opt/mattermost/bin/mattermost``. The install directory is everything before the first occurrence of the string ``/mattermost``. In this example, the ``{install-path}`` is ``/opt``. If that command does not produce any results because your version is older, try ``whereis platform`` instead.

Reviewer 2: `whereis` only looks up standard binary places and $PATH. If e.g. **/opt/mattermost/bin** wasn't added to the PATH `whereis` will return nothing. To find the *mattermost* binary something like `find / -executable -type f -iname mattermost 2> /dev/null` would be more generic/useful.

Author: This is great feedback, @mrckndt! Thank you!

        cd {install-path}
        sudo cp -ra mattermost/ mattermost-back-$(date +'%F-%H-%M')/

7. Remove all files **except** data and custom directories from within the current ``mattermost`` directory. 

>   **What's Preserved?**

Reviewer: ```suggestion
  **What's preserved?**
```

  
  By default, your data directories will be preserved with the following commands:``config``, ``logs``, ``plugins``, ``client/plugins``, and ``data`` (unless you have a different value configured for local storage). Custom directories are any directories that you've added to Mattermost and are not preserved by default. Generally, these are TLS keys or other custom information.

  Run ``ls`` on your Mattermost install directory to identify what default folders exist. If your folders match the structure specified in the following note, you can jump to step 8 below.
      
  **A default Mattermost installation has the following files and directories**:

  .. code-block:: sh

    $ ls /opt/mattermost
    ENTERPRISE-EDITION-LICENSE.txt README.md  client  data   i18n  manifest.txt  prepackaged_plugins
    NOTICE.txt                      bin        config  fonts  logs  plugins       templates

>   **Clear Mattermost Contents**

Reviewer: ```suggestion
  **Clear Mattermost contents**
```

>   The following command clears the contents of mattermost, preserving only those directories and their contents. You should first modify the last part to ``xargs echo rm -r`` to verify what will be executed.

Reviewer: ```suggestion
  The following command clears the contents of the ``mattermost`` folder, preserving only the specified directories and their contents. You should first modify the last part to ``xargs echo rm -r`` to verify what will be executed.
```

>   If you store TLSCert/TLSKey files or other information within your ``/opt/mattermost`` folder, you should add ``-o -path mattermost/yourFolderHere`` to the below command or you will have to manually copy the TLSCert/TLSKey files from the backup into the new install. 

Reviewer: ```suggestion
  If you store TLSCert/TLSKey files or other information within your ``/opt/mattermost`` folder, you should add ``-o -path mattermost/yourFolderHere`` to the below command or you'll have to manually copy the TLSCert/TLSKey files from the backup into the new install.
```

  .. code-block:: sh

    sudo find mattermost/ mattermost/client/ -mindepth 1 -maxdepth 1 \! \( -type d \( -path mattermost/client -o -path mattermost/client/plugins -o -path mattermost/config -o -path mattermost/logs -o -path mattermost/plugins -o -path mattermost/data \) -prune \) | sort | sudo xargs rm -r  
  
  **Using Bleve Search**

>   If using `Bleve Search <https://docs.mattermost.com/deploy/bleve-search.html>`__, the index directory path won't be preserved with the commands below if the directory exists *within* the ``mattermost`` directory. 

Reviewer: ```suggestion
  If using `Bleve Search <https://docs.mattermost.com/deploy/bleve-search.html>`__, and the directory exists *within* the ``mattermost`` directory, the index directory path won't be preserved with the commands below. 
```

  .. code-block:: sh
         
    sudo chown -hR mattermost:mattermost /tmp/mattermost-upgrade/
     
.. note::
    
  - If you didn't use ``mattermost`` as the owner and group of the install directory, run ``sudo chown -hR {owner}:{group} tmp/mattermost-upgrade/``.
  - If you're uncertain what owner or group was defined, use the ``ls -l {install-path}/mattermost/bin/mattermost`` command to obtain them.

9. Copy the new files to your install directory and remove the temporary files.

  .. code-block:: sh

   sudo cp -an /tmp/mattermost-upgrade/. mattermost/
   sudo rm -r /tmp/mattermost-upgrade/
   sudo rm -i /tmp/mattermost*.gz

  .. note::
    The ``n`` (no-clobber) flag and trailing ``.`` on source are very important.

10. If you want to use port 80 to serve your server, or if you have TLS set up on your Mattermost server, you **must** activate the CAP_NET_BIND_SERVICE capability to allow the new Mattermost binary to bind to low ports. For example:

  .. code-block:: sh

    cd {install-path}/mattermost
    sudo setcap cap_net_bind_service=+ep ./bin/mattermost

11. Start your Mattermost server.

  .. code-block:: sh

    sudo systemctl start mattermost

> 12. If you're using a High Availability deployment, you need to apply the steps above on every nodes in your cluster. Once complete, the **Config File MD5** columns in the High Availability section of the system console should be green. If they're yellow, please ensure that all nodes have the same server version and the same configuration.

Reviewer: ```suggestion
12. If you're using a High Availability deployment, you need to apply the steps above on every node in your cluster. Once complete, the **Config File MD5** columns in the High Availability section of the System Console should be green. If they're yellow, please ensure that all nodes have the same server version and the same configuration.
```

        cd {install-path}
        sudo cp -ra mattermost/ mattermost-back-$(date +'%F-%H-%M')/

7. Remove all files **except** data and custom directories from within the current ``mattermost`` directory. 

  **What's preserved?**
  
  By default, your data directories will be preserved with the following commands:``config``, ``logs``, ``plugins``, ``client/plugins``, and ``data`` (unless you have a different value configured for local storage). Custom directories are any directories that you've added to Mattermost and are not preserved by default. Generally, these are TLS keys or other custom information.

  Run ``ls`` on your Mattermost install directory to identify what default folders exist. If your folders match the structure specified in the following note, you can jump to step 8 below.
      
  **A default Mattermost installation has the following files and directories**:

  .. code-block:: sh

    $ ls /opt/mattermost
    ENTERPRISE-EDITION-LICENSE.txt README.md  client  data   i18n  manifest.txt  prepackaged_plugins
    NOTICE.txt                      bin        config  fonts  logs  plugins       templates

  **Clear Mattermost contents**

>   The following command clears the contents of the ``mattermost`` folder, preserving only the specified directories and their contents. You should first modify the last part to ``xargs echo rm -r`` to verify what will be executed.

Reviewer 2: I think it's a good idea to change this to a 'test' run with `sudo find mattermost/ mattermost/client/ -mindepth 1 -maxdepth 1 \! \( -type d \( -path mattermost/client -o -path mattermost/client/plugins -o -path mattermost/config -o -path mattermost/logs -o -path mattermost/plugins -o -path mattermost/data \) -prune \)` instead of the remove command without sudo. In case a user switched to **root** or **mattermost** and forgot about it, `xargs echo rm -r` will remove the data anyway.

  .. code-block:: sh

    sudo find mattermost/ mattermost/client/ -mindepth 1 -maxdepth 1 \! \( -type d \( -path mattermost/client -o -path mattermost/client/plugins -o -path mattermost/config -o -path mattermost/logs -o -path mattermost/plugins -o -path mattermost/data \) -prune \) | sort | sudo xargs rm -r  
  
  **Using Bleve Search**

  If using `Bleve Search <https://docs.mattermost.com/deploy/bleve-search.html>`__, and the directory exists *within* the ``mattermost`` directory, the index directory path won't be preserved with the commands below. 
  
  - You can either move the bleve index directory out from the ``mattermost`` directory before upgrading or, following an upgrade, you can copy the contents of the bleve index directory from the ``backup`` directory. 
  - You can then store that directory or re-index as preferred. 
  - The bleve indexes can be migrated without reindexing between Mattermost versions. See our `Configuration Settings <https://docs.mattermost.com/configure/configuration-settings.html#bleve-settings-experimental>`__ documentation for details on setting the bleve index directory.
      

> 8. Change ownership of the new files before copying them. For example:

Reviewer 2: I would put this before point 11 (right before starting the server). In case the user managed to change the permissions while doing the upgrade (e.g. by copying files/folders with sudo) it's not enough to change the owner and group of just the install files.

  .. code-block:: sh
         
    sudo chown -hR mattermost:mattermost /tmp/mattermost-upgrade/

>      

Reviewer 2: In case this moves down: `sudo chown -R mattermost:mattermost {install-path}`. The `-h` option can be omitted and probably introduces more issues than it solves.

.. note::
    
  - If you didn't use ``mattermost`` as the owner and group of the install directory, run ``sudo chown -hR {owner}:{group} tmp/mattermost-upgrade/``.
  - If you're uncertain what owner or group was defined, use the ``ls -l {install-path}/mattermost/bin/mattermost`` command to obtain them.

9. Copy the new files to your install directory and remove the temporary files.

  .. code-block:: sh

   sudo cp -an /tmp/mattermost-upgrade/. mattermost/
   sudo rm -r /tmp/mattermost-upgrade/
   sudo rm -i /tmp/mattermost*.gz

  .. note::
    The ``n`` (no-clobber) flag and trailing ``.`` on source are very important.

> 

Reviewer 2: *The ``n`` (no-clobber) flag preserves existing configs and logs in your installation path and the trailing ``.`` on source ensures all install files are copied.*? - This makes it more clear and less opaque.

> 10. If you want to use port 80 to serve your server, or if you have TLS set up on your Mattermost server, you **must** activate the CAP_NET_BIND_SERVICE capability to allow the new Mattermost binary to bind to low ports. For example:

Reviewer 2: Maybe add an explicit note on port 443 (instead of or in addition to TLS) as this is a common question which arises sometimes and not everyone knows the terminology/tech behind this. Also low ports isn't really specific and could be changed to *ports lower than 1024*. This leaves less open questions to the users imo.
